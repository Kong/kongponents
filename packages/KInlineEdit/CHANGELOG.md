# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [4.0.1](https://github.com/Kong/kongponents/compare/v4.0.0...v4.0.1) (2021-09-07)

**Note:** Version bump only for package @kongponents/kinlineedit





# 4.0.0 (2021-09-03)


### Features

* **KInlineEdit:** Adds KInlineEdit component ([#222](https://github.com/Kong/kongponents/issues/222)) ([50720fd](https://github.com/Kong/kongponents/commit/50720fd54aa6812bb58aa9920ce9e0d56d9151df))
* **style:** Updates Colors & Component Styles ([#225](https://github.com/Kong/kongponents/issues/225)) ([642a26b](https://github.com/Kong/kongponents/commit/642a26b888cbf8b4abbe32965729d44d54f9dfef))


### BREAKING CHANGES

* **style:** All new colors and color naming. Various component style improvements.





## [1.0.1](https://github.com/Kong/kongponents/compare/@kongponents/kinlineedit@1.0.0...@kongponents/kinlineedit@1.0.1) (2020-09-01)

**Note:** Version bump only for package @kongponents/kinlineedit





# [1.0.0](https://github.com/Kong/kongponents/compare/@kongponents/kinlineedit@0.2.0...@kongponents/kinlineedit@1.0.0) (2020-06-25)


### Features

* **style:** Updates Colors & Component Styles ([#225](https://github.com/Kong/kongponents/issues/225)) ([642a26b](https://github.com/Kong/kongponents/commit/642a26b888cbf8b4abbe32965729d44d54f9dfef))


### BREAKING CHANGES

* **style:** All new colors and color naming. Various component style improvements.





# 0.2.0 (2020-06-17)


### Features

* **KInlineEdit:** Adds KInlineEdit component ([#222](https://github.com/Kong/kongponents/issues/222)) ([50720fd](https://github.com/Kong/kongponents/commit/50720fd54aa6812bb58aa9920ce9e0d56d9151df))
