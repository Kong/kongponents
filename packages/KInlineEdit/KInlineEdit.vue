<template>
  <div
    id="editable-wrapper"
    class="k-inline-edit">
    <input
      v-if="isEditing"
      ref="input"
      :style="styles"
      v-model="inputText"
      class="k-input"
      @blur="handleSave"
      @keyup.enter="$event.target.blur()">
    <div
      v-if="!isEditing"
      id="element-content-wrapper"
      @click="handleClick">
      <slot v-if="!isEditing" />
    </div>
  </div>
</template>

<script>
// Styles we want copied from the element
const STYLES = {
  fontSize: 'font-size',
  fontWeight: 'font-weight',
  fontFamily: 'font-family',
  color: 'color',
  margin: 'margin',
  padding: 'padding'
}

export default {
  name: 'KInlineEdit',
  props: {
    /**
     * Sets whether the initial value passed in should be ignored.
     * Useful for times when you have placeholder text and don't want it passed
     * in as a value
     */
    ignoreValue: {
      type: Boolean,
      default: false
    },
    /**
     * An object of css styles to override those plucked from the slot element.
     * Useful if you are styling the default content differently than how the
     * input should display
     */
    styleOverrides: {
      type: Object,
      default: () => ({})
    }
  },

  data () {
    return {
      isEditing: false,
      inputText: '',
      styles: {}
    }
  },

  mounted () {
    if (this.$slots.default[0].tag === undefined) {
      console.warn(`KInlineEdit expects slotted HTML tag.

Example usage:

<KInlineEdit>
  <p>Some text</p>
</KInlineEdit>
      `)
    }
  },

  methods: {
    async handleClick (e) {
      // If clicking the slot wrapper lets exit out as to not
      // copy its styles to the input
      if (e.target.id === 'element-content-wrapper') return

      // Get current STYLES off of the element
      this.styles = { ...this.getStyles(e.target), ...this.styleOverrides }
      this.inputText = this.ignoreValue ? '' : e.target.textContent
      this.isEditing = true

      // Wait for vue to update styles & text
      await this.$nextTick()
      this.$refs.input.focus()
    },

    handleSave (e) {
      this.isEditing = false
      this.$emit('changed', this.inputText)
    },

    getStyles (element) {
      const elementStyles = getComputedStyle(element)

      return Object.keys(STYLES).reduce((acc, cur) => {
        acc[cur] = elementStyles.getPropertyValue(STYLES[cur])

        return acc
      }, {})
    }
  }
}
</script>

<style lang="scss" scoped>
@import '~@kongponents/styles/_variables.scss';
@import '~@kongponents/styles/forms/_inputs.scss';

.k-inline-edit {
  --padding: var(--spacing-xxs) var(--spacing-xs);
  box-sizing: border-box;
  > div {
    display: inline-flex;
    width: 100%;
    cursor: text;
    > * {
      width: 100%;
      border: 1px solid transparent;
      border-radius: 3px;
      padding: var(--padding);
      margin-left: calc(-1 * var(--spacing-xs)); // align the left side of content
      line-height: 1.25;
      overflow: hidden;
      transition: background-color 200ms ease;
    }
    &:hover > * {
      background-color: var(--blue-100);
    }
  }
  .k-input {
    display: inline-flex;
    width: 100%;
    padding: var(--padding);
    &:focus,
    &:hover {
      background-color: var(--white);
    }
  }
}
</style>
