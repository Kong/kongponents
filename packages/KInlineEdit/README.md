# @kongponents/kinlineedit

[![](https://img.shields.io/npm/v/@kongponents/kinlineedit.svg?style=flat-square)](https://www.npmjs.com/package/@kongponents/kinlineedit)

```vue
<template>
  <EditableInput @changed="val => textVal = val">
    <h1>{{ textVal }}</h1>
  </EditableInput>
</template>
<script>
import EditableInput from "@/components/EditableInput";
export default {
  name: "App",
  components: { EditableInput },
  data() {
    return {
      textVal: "Inline Text"
    };
  }
};
</script>
```
