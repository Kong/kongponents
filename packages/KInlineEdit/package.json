{
  "name": "@kongponents/kinlineedit",
  "version": "6.24.14",
  "description": "Inline editable field.",
  "main": "dist/KInlineEdit.umd.min.js",
  "componentName": "KInlineEdit",
  "source": "KInlineEdit.vue",
  "files": [
    "dist",
    "KInlineEdit.vue"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@kongponents/styles": "^6.24.14"
  }
}
