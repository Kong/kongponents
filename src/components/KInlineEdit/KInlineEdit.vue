<template>
  <div
    :id="`editable-wrapper-${inputUuid}`"
    class="k-inline-edit"
  >
    <input
      v-if="isEditing"
      ref="input"
      v-model.trim="inputText"
      :style="styles"
      class="k-input"
      @blur="handleSave"
      @keyup.enter="onEnter"
    >
    <div
      v-if="!isEditing"
      :id="`element-content-wrapper-${inputUuid}`"
      @click="handleClick"
    >
      <slot v-if="!isEditing" />
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, ref, computed, onMounted, nextTick } from 'vue'
import { v1 as uuidv1 } from 'uuid'

// Styles we want copied from the element
const STYLES = {
  fontSize: 'font-size',
  fontWeight: 'font-weight',
  fontFamily: 'font-family',
  color: 'color',
  margin: 'margin',
  padding: 'padding',
}

export default defineComponent({
  name: 'KInlineEdit',
  props: {
    /**
     * Sets whether the initial value passed in should be ignored.
     * Useful for times when you have placeholder text and don't want it passed
     * in as a value
     */
    ignoreValue: {
      type: Boolean,
      default: false,
    },
    /**
     * An object of css styles to override those plucked from the slot element.
     * Useful if you are styling the default content differently than how the
     * input should display
     */
    styleOverrides: {
      type: Object,
      default: () => ({}),
    },
  },
  emits: ['changed'],
  setup(props, { emit, slots }) {
    const input = ref(null)
    const inputUuid = computed((): string => 'editable-wrapper-' + uuidv1())
    const isEditing = ref(false)
    const inputText = ref('')
    const styles = ref({})

    const handleClick = async (e: any): Promise<void> => {
      // If clicking the slot wrapper lets exit out as to not
      // copy its styles to the input
      if (e.target.id === 'element-content-wrapper') return

      // Get current STYLES off of the element
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      styles.value = { ...getStyles(e.target), ...props.styleOverrides }
      inputText.value = props.ignoreValue ? '' : e.target.textContent
      isEditing.value = true

      // Wait for Vue to update styles & text
      await nextTick()

      if (input.value) {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        input.value.focus()
      }
    }

    const handleSave = (): void => {
      isEditing.value = false
      emit('changed', inputText.value)
    }

    const getStyles = (element: HTMLElement) => {
      const elementStyles = getComputedStyle(element)

      return Object.keys(STYLES).reduce((acc: string, cur: string) => {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        acc[cur] = elementStyles.getPropertyValue(STYLES[cur])

        return acc
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
      }, {})
    }

    const onEnter = (e: any): void => {
      e?.target?.blur()
    }

    onMounted(() => {
      try {
        if (!slots.default) {
          throw new Error('KInlineEdit expects a slotted HTML tag.')
        }
      } catch (_) {
        console.error(`KInlineEdit expects a slotted HTML tag.

    Example usage:

      <KInlineEdit>
        <p>Some text</p>
        ^^------add slotted tag
      </KInlineEdit>
    `)
      }
    })

    return {
      input,
      inputUuid,
      isEditing,
      inputText,
      styles,
      handleClick,
      handleSave,
      onEnter,
    }
  },
})
</script>

<style lang="scss" scoped>
@import '@/styles/variables';
@import '@/styles/functions';

.k-inline-edit {
  --padding: var(--spacing-xxs) var(--spacing-xs);
  box-sizing: border-box;

  > :deep(div) {
    display: inline-flex;
    width: 100%;
    cursor: text;

    > * {
      width: 100%;
      border: 1px solid transparent;
      border-radius: 3px;
      padding: var(--padding);
      margin-left: calc(-1 * var(--spacing-xs)); // align the left side of content
      line-height: 1.25;
      overflow: hidden;
      transition: background-color 200ms ease;
    }

    &:hover > * {
      background-color: var(--grey-200);
    }
  }

  .k-input {
    display: inline-flex;
    width: 100%;
    padding: var(--padding);

    &:focus,
    &:hover {
      background-color: var(--white);
    }
  }
}
</style>
