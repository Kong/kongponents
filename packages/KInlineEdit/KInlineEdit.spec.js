import { mount } from '@vue/test-utils'
import KInlineEdit from '@/KInlineEdit/KInlineEdit'

describe('KInlineEdit', () => {
  it('copies element css to input on click', () => {
    const wrapper = mount(KInlineEdit, {
      slots: { default: '<p style="padding:10px">text</p>' }
    })

    wrapper.find('p').trigger('click')

    expect(wrapper.find('.k-input').attributes().style).toBe('padding: 10px;')
  })

  it('console warns when html tag not passed in slot', () => {
    console.warn = jest.fn()
    mount(KInlineEdit, {
      slots: { default: 'text' }
    })

    expect(console.warn).toBeTruthy()
  })

  it('emits updated text on blur', async () => {
    const inputText = 'test'
    const wrapper = mount(KInlineEdit, {
      slots: { default: '<p>text</p>' }
    })

    wrapper.find('p').trigger('click')

    await wrapper.vm.$nextTick()

    const input = wrapper.find('.k-input')

    input.setValue(inputText)
    input.trigger('blur')
    expect(wrapper.emitted().changed).toBeTruthy()
    expect(wrapper.emitted().changed[0][0]).toEqual(inputText)
  })

  it('matches snapshot', () => {
    const wrapper = mount(KInlineEdit, {
      slots: { default: '<p>text</p>' }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
