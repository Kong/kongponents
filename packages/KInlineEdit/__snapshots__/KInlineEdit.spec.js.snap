// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KInlineEdit matches snapshot 1`] = `
<div id="editable-wrapper" class="k-inline-edit">
  <!---->
  <div id="element-content-wrapper">
    <p>text</p>
  </div>
</div>
`;
